### DISCUSSION TITLE
GA of Chiseled Ubuntu Container

### DISCUSSION THREAD

**Discussion by gareth-houghton** [dotnet/dotnet-docker#4137] [Q&A] [answered]
I'm not sure where is best to ask this question so I'm happy to be pointed in a different direction. 

It was great to see the [announcement](https://devblogs.microsoft.com/dotnet/dotnet-6-is-now-in-ubuntu-2204/) of .NET 6 in Chiseled Ubuntu containers back in August. The announcement mentions that:

> Note: The images will be offered in our nightly repos while the chiseled offering is in preview. We’ll make another announcement when they are supported in production. It will be sometime this year, but we haven’t picked a timeframe, since we’ve been focused on basic enablement.

I'm interested in knowing if there is an up to date view on when this will be ready for GA. It'd be good to get a feel for when we can expect chiseled images to be in the main .NET registries.

**Comment by mthalman** [comment, marked as answer]
Thanks for your interest in the chiseled containers. We don't yet have a timeframe for when this will be ready as a GA release. We continue to meet regularly with our friends at Canonical on this matter. One thing that we're working on is ensuring there's the right level of technical support between our two companies for the whole stack that's involved here. We'll keep this thread informed as we have more information.

**Reply by gareth-houghton** [reply]
Hey @mthalman, thanks for your quick response. It's great to hear that it's still on the cards, is it looking like it will be this year or is it still too early for that?

**Reply by mthalman** [reply]
Still too early. We're hoping that by next month we'll have an idea of a timeline.

**Reply by gareth-houghton** [reply]
That's great, thanks for the clarity here. I'll keep an eye on this thread for some news! Thanks for your help @mthalman

**Reply by orjan** [reply]
Are there any news when to expect a stable version of the chiseled containers?

**Reply by mthalman** [reply]
> Are there any news when to expect a stable version of the chiseled containers?

Not at this time. We're still coordinating with our counterparts at Canonical on support for chiseled containers.

**Reply by GustavoTorricoTR** [reply]
any news about, when stable versions will be offered?

**Reply by richlander** [reply]
Same answer as before, sadly. We're still waiting on Canonical to publish a date. We remain enthusiastic about this image type and of publishing them. We announced today that they we are going live with these images for .NET 8 including dotnet-monitor exclusively publishing with these images (for .NET 8). Our hope is that we'll go live with them for .NET 6 and 7 ahead of that.

**Reply by mthalman** [reply]
See our .NET 8 Preview 1 announcement that highlights the chiseled containers here: https://github.com/dotnet/dotnet-docker/discussions/4440

**Reply by richlander** [reply]
The underlying reason for the delay is waiting for a way to produce SBOMs for Chiseled images so that they can be audited by a tool like syft.

**Reply by 64J0** [reply]
Any update here?

**Comment by ptr1120** [comment]
Hello,

I am also very interested in the question of GA of Chiseled Ubuntu Container.

Currently, I see 3 different approaches for a distroless Dotnet runtime images:
- [Ubuntu Chiseled](https://github.com/ubuntu-rocks/dotnet/)
- [CBL-Mariner](https://github.com/microsoft/CBL-Mariner)
- [Investigate production of distroless-style Alpine images](https://github.com/dotnet/dotnet-docker/issues/4162)

The Chiseled Ubuntu Container fits best for me because it has less stuff inside (in contrast to CBL-Mariner) and it provides a glibc compatible environment (in contrast to alpine musl linux).

Since this topic is critical in terms of security and performance it would be important to get a statement on whether or not the Chiseled Ubuntu Container will be the officially supported way to produce distroless Dotnet runtime images in the future.

**Comment by mthalman** [comment]
We've posted an update about our Chiseled container images: https://github.com/dotnet/dotnet-docker/discussions/4739. Please take a look and provide feedback.

**Reply by 64J0** [reply]
Nice!

**Comment by mthalman** [comment]
For those that are interested in Visual Studio debugger support for distoless containers, please provide any feedback you have to this post: https://developercommunity.visualstudio.com/t/Debugger-should-support-attaching-to-pro/10461836

**Reply by nhart12** [reply]
@mthalman  I was able to get debugging working from visual studio by chiseling out the busybox_bins and adding them to my fastmode stage


```
FROM golang:1.20 AS chisel
RUN go install github.com/canonical/chisel/cmd/chisel@latest

FROM <build-SDK-stage> as <debugdependencies-layer> 
....
COPY --from=chisel /go/bin/chisel /bin/
RUN mkdir /debug-tools && chisel cut --root /debug-tools busybox_bins 
RUN ["/debug-tools/bin/busybox", "--install", "/debug-tools/bin"] #add all busybox links


FROM <base-stage> as <whatever-name-of-your-DockerfileFastmodeStage>
COPY --from=dependencies /debug-tools /
```

**Comment by aygalinc** [comment]
Any plan on the GA for .net 8 & .net 7 ?

**Reply by mthalman** [reply]
The plan is to provide GA support for chiseled containers next month. This would apply to all supported .NET versions (6, 7, 8).

**Comment by mthalman** [comment]
Hey folks! GA support for Ubuntu Chiseled containers is now here for .NET 6, 7, and 8. You can grab the images today:

```
docker pull mcr.microsoft.com/dotnet/aspnet:6.0-jammy-chiseled
docker pull mcr.microsoft.com/dotnet/aspnet:7.0-jammy-chiseled
docker pull mcr.microsoft.com/dotnet/aspnet:8.0-jammy-chiseled
```

See more info at https://github.com/dotnet/dotnet-docker/blob/main/documentation/ubuntu-chiseled.md.
